pinner/pinner.c

```diff
@@ -0,0 +1,187 @@
+
+	g_slist_free_full(list, g_free);
+	container->list = NULL;
+}
+
+
+/* A function adapted (pinched) from filebrowser.c */
+static GtkWidget *create_popup_menu(gpointer pdata)
+{
+	GtkWidget *item, *menu;
+
+	menu = gtk_menu_new();
+
+	/* 3? What should the size be? */
+	/* https://docs.gtk.org/gtk3/ctor.Image.new_from_icon_name.html */
+	item = gtk_image_new_from_icon_name("edit-clear", 3);
+	gtk_widget_show(item);
+	gtk_container_add(GTK_CONTAINER(menu), item);
+	g_signal_connect(item, "activate", G_CALLBACK(slist_free_wrapper), pdata);
+
+	return menu;
+}
+
+bool is_duplicate(GSList *list, const gchar* file_name)
+{
+	GSList *iter;
+	for (iter = list; iter != NULL; iter = g_slist_next(iter)) {
+		if (g_strcmp0((const gchar *)iter->data, file_name) == 0) {
+			/* We'll probably want to alert the user the document already
+			 * is pinned */
+			return true;
+		}
+	}
+	return false;
+}
+
+static void pin_activate_cb(GtkMenuItem *menuitem, gpointer pdata)
+{
+	struct pindata *container = pdata;
+	GeanyPlugin *plugin = container->plugin;
+	GSList *list = container->list;
+
+	GeanyDocument *doc = document_get_current();
+	if (doc == NULL)
+		return;
+
+	// DEBUG: Print the elements in the list
+	//printf("List elements:\n");
+	//for (GSList *iter = list; iter != NULL; iter = g_slist_next(iter))
+		//printf("%s\n", (gchar *)iter->data);
+
+	if (is_duplicate(list, doc->file_name))
+		return;
+
+	/* This must be freed when nodes are removed from the list */
+	gchar *tmp_file_name = g_strdup(doc->file_name);
+
+	container->list = g_slist_append(list, tmp_file_name);
+	GtkWidget *label = gtk_label_new_with_mnemonic(doc->file_name);
+	gtk_widget_show(label);
+	gtk_box_pack_start(GTK_BOX(pinned_view_vbox), label, FALSE, FALSE, 0);
+	gtk_notebook_set_current_page(GTK_NOTEBOOK(plugin->geany_data->main_widgets->sidebar_notebook), page_number);
+
+	return;
+}
+
+static gboolean on_button_press(GtkWidget *widget, GdkEventButton *event, gpointer pdata)
+{
+	//if (event->button == 1 && event->type == GDK_2BUTTON_PRESS)
+	//{
+		//on_open_clicked(NULL, NULL);
+		//return TRUE;
+	//}
+	if (event->button == 3)
+	{
+		static GtkWidget *popup_menu = NULL;
+
+		if (popup_menu == NULL)
+			popup_menu = create_popup_menu(pdata);
+
+		gtk_menu_popup_at_pointer(GTK_MENU(popup_menu), (GdkEvent *) event);
+		/* don't return TRUE here, unless the selection won't be changed */
+	}
+	return FALSE;
+}
+
+static gboolean pin_init(GeanyPlugin *plugin, gpointer pdata)
+{
+	GtkWidget *main_menu_item;
+
+	struct pindata *container = init_pindata();
+	container->plugin = plugin;
+
+	// Create a new menu item and show it
+	main_menu_item = gtk_menu_item_new_with_mnemonic("Pin Document");
+	gtk_widget_show(main_menu_item);
+	gtk_container_add(GTK_CONTAINER(plugin->geany_data->main_widgets->tools_menu),
+		main_menu_item);
+	g_signal_connect(main_menu_item, "activate",
+		G_CALLBACK(pin_activate_cb), container);
+	g_signal_connect(pinned_view_vbox, "button-press-event",
```

Comment:
No popup menu appears and I see in the console:

```
(geany:1064826): GLib-GObject-CRITICAL **: 10:20:24.486: invalid (NULL) pointer instance

(geany:1064826): GLib-GObject-CRITICAL **: 10:20:24.486: g_signal_connect_data: assertion 'G_TYPE_CHECK_INSTANCE (instance)' failed
```

I think using 'pinned_view_vbox' as the 2nd argument  here is not correct.

I get no compiler warnings.

Reply:
Fixed.